docs/groovy/conceptual/positional-index-safety.md

```diff
@@ -0,0 +1,334 @@
+- **`k`** is the row key, a unique identifier for each row that may differ from positional index.
+- **Column array variables** allow direct access to column data arrays, which become invalid when table structure changes.
+
+## Key terms
+
+Before diving into the details, let's define the key concepts:
+
+- **Row key (`k`)**: A unique identifier assigned to each row in a table. Row keys are used internally by the engine to track rows and are typically sequential integers starting from 0.
+
+- **Positional index (`i`)**: The current position of a row within a table, starting from 0 for the first row. Unlike row keys, positional indices always reflect the current order of rows.
+
+- **Shifts**: When a table grows beyond its pre-allocated row key space, the engine may need to move (shift) other tables' rows to new row key positions to make room. The row data itself doesn't change, but the row keys do.
+
+- **Constituent tables**: The individual tables that are being merged together. Each constituent table gets its own allocated range of row keys in the merged result.
+
+- **Row key slots**: Pre-allocated ranges of row keys reserved for each constituent table. For example, table A might get slots 0-4095, while table B gets slots 4096-8191.
+
+- **Refreshing tables**: Tables that can change over time (live data). This includes tables created from live data sources (like `timeTable`, Kafka streams, database connections), tables that are the result of operations on refreshing tables, and any table where rows can be added, removed, or shifted after creation. In contrast, **static tables** never change after creation - they contain fixed data that remains constant.
+
+- **Append-only tables**: A special type of refreshing table where new rows are only added at the end (never inserted, moved, or deleted), making positional indices safe to use.
+
+## Why positional indices become unsafe
+
+When a refreshing table changes, several things can happen that make positional indices incorrect:
+
+### 1. Row additions
+
+When new rows are added to a table, the positional indices of existing rows may change. Consider this example:
+
+```groovy order=null
+// Start with a simple live table
+liveTable = timeTable("PT1s").update("value = randomInt(1, 100)")
+
+// If we could add a positional index formula (this would fail):
+// indexedTable = liveTable.update("position = i", "indexed_value = value + i")
+```
+
+**The problem:** As new rows are added every second, existing rows' positional indices shift:
+
+- At T=0: Row with `value=50` is at `i=0`, so `indexed_value` would be `50 + 0 = 50`
+- At T=1: New row added at top, previous row moves to `i=1`, but `indexed_value` stays `50`
+- At T=2: Previous row now at `i=2`, but `indexed_value` still shows `50` instead of `52`
+
+### 2. Row removals
+
+When rows are removed, positional indices of remaining rows change. Here's a concrete example:
+
+```groovy order=null
```

Comment:
Like the code block above it, this runs just fine even if you un-comment the last line.